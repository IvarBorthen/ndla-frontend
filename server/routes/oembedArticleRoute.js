/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { matchPath } from 'react-router-dom';
import { BAD_REQUEST, INTERNAL_SERVER_ERROR } from 'http-status';
import parseUrl from 'parse-url';
import { isValidLocale } from '../../src/i18n';
import { fetchArticle } from '../../src/containers/ArticlePage/articleApi';
import { articlePath } from '../../src/routes';
import config from '../../src/config';

export function parseAndMatchUrl(url) {
  const { pathname } = parseUrl(url);
  const paths = pathname.split('/');
  if (isValidLocale(paths[1])) {
    return matchPath(pathname, `/:lang${articlePath}`);
  }
  return matchPath(pathname, articlePath);
}

export async function oembedArticleRoute(req) {
  const { url } = req.query;
  if (!url) {
    return {
      status: BAD_REQUEST,
      data: 'Bad request. Missing url param.',
    };
  }

  const match = parseAndMatchUrl(url);

  if (!match) {
    return {
      status: BAD_REQUEST,
      data: 'Bad request. Invalid url.',
    };
  }

  const {
    params: { articleId, plainResourceId: resourceId, lang = 'nb' },
  } = match;

  try {
    const article = await fetchArticle(articleId, lang);
    return {
      data: {
        type: 'rich',
        version: '1.0', // oEmbed version
        height: req.query.height ? req.query.height : 800,
        width: req.query.width ? req.query.width : 800,
        title: article.title,
        html: `<iframe src="${
          config.ndlaFrontendDomain
        }/article-iframe/${lang}/${resourceId}/${articleId}" frameborder="0" />`,
      },
    };
  } catch (error) {
    console.error(error);
    const status = error.status || INTERNAL_SERVER_ERROR;
    return {
      status,
      data: 'Internal server error',
    };
  }
}
