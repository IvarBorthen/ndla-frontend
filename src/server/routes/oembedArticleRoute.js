/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { matchPath } from 'react-router-dom';
import { BAD_REQUEST, INTERNAL_SERVER_ERROR } from 'http-status';
import parseUrl from 'parse-url';
import { isValidLocale } from '../../i18n';
import { getArticleIdFromResource } from '../../containers/Resources/resourceHelpers';
import { fetchResource } from '../../containers/Resources/resourceApi';
import { articlePath } from '../../routes';
import config from '../../config';

const log = require('../../util/logger');

export function parseAndMatchUrl(url) {
  const { pathname } = parseUrl(url);
  const paths = pathname.split('/');
  if (isValidLocale(paths[1])) {
    return matchPath(pathname, `/:lang${articlePath}`);
  }
  return matchPath(pathname, articlePath);
}

export async function oembedArticleRoute(req) {
  const { url } = req.query;
  if (!url) {
    return {
      status: BAD_REQUEST,
      data: 'Bad request. Missing url param.',
    };
  }

  const match = parseAndMatchUrl(url);

  if (!match) {
    return {
      status: BAD_REQUEST,
      data: 'Bad request. Invalid url.',
    };
  }

  const { params: { resourceId, lang = 'nb' } } = match;

  try {
    const resource = await fetchResource(`urn:resource:${resourceId}`, lang);
    const articleId = getArticleIdFromResource(resource);
    return {
      data: {
        type: 'rich',
        version: '1.0', // oEmbed version
        height: req.query.height ? req.query.height : 800,
        width: req.query.width ? req.query.width : 800,
        title: resource.title,
        html: `<iframe src="${
          config.ndlaFrontendDomain
        }/article-iframe/${lang}/${
          resource.id
        }/${articleId}" frameborder="0" />`,
      },
    };
  } catch (error) {
    log.error(error);
    const status = error.status || INTERNAL_SERVER_ERROR;
    return {
      status,
      data: 'Internal server error',
    };
  }
}
