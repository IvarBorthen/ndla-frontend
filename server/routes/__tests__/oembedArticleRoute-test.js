/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import nock from 'nock';
import sinon from 'sinon';
import { parseAndMatchUrl, oembedArticleRoute } from '../oembedArticleRoute';

const response = {
  setHeader: jest.fn(),
  status: jest.fn(),
  json: jest.fn(),
};
response.status.mockImplementation(() => response);

const validArticleUrl1 =
  'https://ndla-frontend.test.api.ndla.no/article/urn:subject:3/urn:topic:1:55163/urn:topic:1:168398/urn:resource:1:1682/26050';
const validArticleUrl2 =
  'http://localhost:3000/article/urn:subject:3/urn:topic:1:168542/urn:topic:1:173292/urn:resource:1:168554/127';

const validArticleUrlWithLang =
  'https://ndla-frontend.test.api.ndla.no/nn/article/urn:subject:3/urn:topic:1:55163/urn:topic:1:168398/urn:resource:1:1682/26050';
const unvalidArticleUrl =
  'https://ndla-frontend.test.api.ndla.no/subjects/urn:subject:3/urn:topic:1:55163';

test('parseAndMatchUrl', () => {
  expect(parseAndMatchUrl(validArticleUrl1)).toMatchSnapshot();
  expect(parseAndMatchUrl(validArticleUrl2)).toMatchSnapshot();
  expect(parseAndMatchUrl(validArticleUrlWithLang)).toMatchSnapshot();
  expect(parseAndMatchUrl(unvalidArticleUrl)).toBe(null);
});

test('oembedArticleRoute success', async () => {
  nock('http://ndla-api')
    .get('/article-converter/json/nb/26050')
    .reply(200, { id: 123, title: 'unit test', metaData: {} });

  const spy = sinon.spy(response, 'json');
  await oembedArticleRoute(
    {
      query: {
        url: validArticleUrl1,
      },
    },
    response,
    { access_token: '123' },
  );

  expect(spy.calledOnce).toBe(true);
  expect(spy.getCall(0).args[0]).toMatchSnapshot();
  spy.restore();
});

test('oembedArticleRoute invalid url', async () => {
  const spy = sinon.spy(response, 'json');

  await oembedArticleRoute(
    {
      query: {
        url: unvalidArticleUrl,
      },
    },
    response,
    { access_token: '123' },
  );

  expect(spy.calledOnce).toBe(true);
  expect(spy.getCall(0).args[0]).toMatchSnapshot();
  spy.restore();
});
