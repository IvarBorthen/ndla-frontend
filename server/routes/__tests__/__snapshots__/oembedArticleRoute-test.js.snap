// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oembedArticleRoute invalid url 1`] = `
Object {
  "status": 400,
  "text": "Bad request. Invalid url.",
}
`;

exports[`oembedArticleRoute success 1`] = `
Object {
  "height": 800,
  "html": "<iframe src=\\"https://ndla-frontend.test.api.ndla.no/article-iframe/nb/1:1682/26050\\" frameborder=\\"0\\" />",
  "title": "unit test",
  "type": "rich",
  "version": "1.0",
  "width": 800,
}
`;

exports[`parseAndMatchUrl 1`] = `
Object {
  "isExact": true,
  "params": Object {
    "0": "urn:topic:1:168542",
    "articleId": "127",
    "plainResourceId": "1:168554",
    "subjectId": "urn:subject:3",
    "topicId": "urn:topic:1:173292",
  },
  "path": "/article/:subjectId/(.*)/:topicId/urn\\\\:resource\\\\::plainResourceId/:articleId",
  "url": "/article/urn:subject:3/urn:topic:1:168542/urn:topic:1:173292/urn:resource:1:168554/127",
}
`;

exports[`parseAndMatchUrl 2`] = `
Object {
  "isExact": true,
  "params": Object {
    "0": "urn:topic:1:55163",
    "articleId": "26050",
    "plainResourceId": "1:1682",
    "subjectId": "urn:subject:3",
    "topicId": "urn:topic:1:168398",
  },
  "path": "/article/:subjectId/(.*)/:topicId/urn\\\\:resource\\\\::plainResourceId/:articleId",
  "url": "/article/urn:subject:3/urn:topic:1:55163/urn:topic:1:168398/urn:resource:1:1682/26050",
}
`;

exports[`parseAndMatchUrl 3`] = `
Object {
  "isExact": true,
  "params": Object {
    "0": "urn:topic:1:55163",
    "articleId": "26050",
    "lang": "nn",
    "plainResourceId": "1:1682",
    "subjectId": "urn:subject:3",
    "topicId": "urn:topic:1:168398",
  },
  "path": "/:lang/article/:subjectId/(.*)/:topicId/urn\\\\:resource\\\\::plainResourceId/:articleId",
  "url": "/nn/article/urn:subject:3/urn:topic:1:55163/urn:topic:1:168398/urn:resource:1:1682/26050",
}
`;
